File: Proyecto Fin de Curso Figueroa/Backend/src/main/java/uteq/edu/ec/artisync/service/seguridad/RolePermissionService.java
package uteq.edu.ec.artisync.service.seguridad;
import uteq.edu.ec.artisync.repository.seguridad.*;
import uteq.edu.ec.artisync.repository.perfil.*;

import uteq.edu.ec.artisync.dto.seguridad.request.CreateRoleRequest;
import uteq.edu.ec.artisync.dto.seguridad.request.UpdateRoleRequest;
import uteq.edu.ec.artisync.dto.seguridad.response.PermisoResponse;
import uteq.edu.ec.artisync.dto.seguridad.response.RolResponse;

import java.util.List;

public interface RolePermissionService {

    List<RolResponse> getAllRoles();

    List<PermisoResponse> getAllPermisos();

    List<String> getPermissionsByRole(String roleName);

    void syncPermissions(String roleName, List<String> permissionCodes);

    RolResponse createRole(CreateRoleRequest request);

    RolResponse updateRole(Long idRol, UpdateRoleRequest request);

    void deleteRole(Long idRol);
}

File: Proyecto Fin de Curso Figueroa/Frontend/src/app/services/role-permission.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { MessageResponse } from '../models/auth.model';

export interface RolMatrixResponse {
  idRol: number;
  nombreRol: string;
  descripcionRol: string;
  permisos: string[];
}

export interface PermisoCatalogResponse {
  idPermiso: number;
  nombrePermiso: string;
  moduloAplicacion: string;
}

export interface SyncPermissionsRequest {
  roleName: string;
  permissionCodes: string[];
}

export interface CreateRoleRequest {
  nombreRol: string;
  descripcionRol?: string;
  permisosIniciales?: string[];
}

export interface UpdateRoleRequest {
  descripcionRol?: string;
}

@Injectable({
  providedIn: 'root'
})
export class RolePermissionService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/admin/role-permissions`;

  getAllRoles(): Observable<RolMatrixResponse[]> {
    return this.http.get<RolMatrixResponse[]>(`${this.apiUrl}/roles`);
  }

  getAllPermisos(): Observable<PermisoCatalogResponse[]> {
    return this.http.get<PermisoCatalogResponse[]>(`${this.apiUrl}/permisos`);
  }

  getPermissionsByRole(roleName: string): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/${roleName}`);
  }

  syncPermissions(request: SyncPermissionsRequest): Observable<MessageResponse> {
    return this.http.put<MessageResponse>(`${this.apiUrl}/sync`, request);
  }

  createRole(request: CreateRoleRequest): Observable<RolMatrixResponse> {
    return this.http.post<RolMatrixResponse>(`${this.apiUrl}/roles`, request);
  }

  updateRole(idRol: number, request: UpdateRoleRequest): Observable<RolMatrixResponse> {
    return this.http.put<RolMatrixResponse>(`${this.apiUrl}/roles/${idRol}`, request);
  }

  deleteRole(idRol: number): Observable<MessageResponse> {
    return this.http.delete<MessageResponse>(`${this.apiUrl}/roles/${idRol}`);
  }
}
